Replay Generic Inbound API

esoergel opened the pull request:
## Product Description
Adds a retry button to the log details page
![image](https://user-images.githubusercontent.com/2367539/206298791-0d5cd52a-7fc2-475d-b5d6-be4904b6ae4e.png)

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-2414
https://docs.google.com/document/d/1lpGyVzur2G2lFOQJJ2nNiNQzasZFMccJiFtRxS5CSPo/edit#
This PR adds a function that accepts a log of an existing request and reprocesses it.

To do this, the PR first introduces two intermediary classes, `RequestData` and `ApiResponse`.  These exist as abstractions over the HTTP request and responses, but they don't need to operate in the context of an actual django web request.  The design goal was to make the main business logic entry point (`execute_generic_api`) accept the one and return the other, so it could be called without a real request, and so its response could be interpreted in a domain-specific context.

This is then minimally hooked in to the UI - longer term I'd like to see this integrated into the report, including support for bulk actions.

## Feature Flag
Generic inbound APIs

## Safety Assurance

### Safety story
pre-release feature

### Automated test coverage

Included in PR

### QA Plan

N/A

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Wrap request params in a custom dataclass

This will let us route initial requests and reprocessed requests through
the same code

esoergel committed:
Create RequestData from log and verify equality

esoergel committed:
Make only "standard" headers available to the API

This means we can maintain parity between original requests and retries,
plus it seems generally safer to exclude say, the authentication header

esoergel committed:
This always uses a list

Since `data` is always a list, `handle_case_update` always returns cases
as a list, even if there's only one there.  This also makes it easier to
do what I need to do next, though I think we can bring back the single
case response format by inspecting the inbound data later if desirable

esoergel committed:
revisit these fields

esoergel committed:
Use intermediary response class

esoergel committed:
Move error processing into core

esoergel committed:
Move status code parsing to class

esoergel committed:
Make functional retry backend

esoergel commented on corehq/motech/generic_inbound/views.py:
```diff
@@ -242,13 +242,14 @@ def _log_api_request(api, request, response):
         status = RequestLog.Status.ERROR
 
-    response_body = response.content.decode('utf-8')
+    # TODO, revisit this
```
Looking for input here (perhaps best to revisit this comment after reviewing the rest of the PR).  Given that the "response" is JSON, does it make sense to store the raw response body?  At the moment it's a deterministic transform of the response JSON.  I understand in the future we might want to include a response transform function, in which case storing before and after that does make sense, but it feels vague enough that I'm not sure it make sense to sorta implement it now.

snopoke replied:
> I understand in the future we might want to include a response transform function, in which case storing before and after that does make sense, but it feels vague enough that I'm not sure it make sense to sorta implement it now.

I agree that it doesn't make a lot of sense right now. Fine to drop it or at least remove it's display from the UI since it is a bit confusing. I was also thinking of renaming it in the UI to clarify the difference:
`raw response` -> `Case API response`
`response body` -> `Response to Client`

esoergel commented on corehq/motech/generic_inbound/views.py:
```diff
@@ -242,13 +242,14 @@ def _log_api_request(api, request, response):
         status=status,
         response_status=response.status_code,
-        error_message=response_body if not is_success else '',
+        # error_message=response_body if not is_success else '',
```
At the moment this is a string like `'{"error": "validation error", "errors": ["Invalid request"]}'`, should we just store the value associated with the `error` key when present?  What was the goal of storing this again (we have the full response stored in the processing attempt)?  Guessing we would want to put this as a column in the list view, but the current version doesn't have that (only `status`).

snopoke replied:
I think the intent was as you suggest, to display it in the report but it doesn't seem strictly necessary as long as that data is available in the processing attempt.

esoergel replied:
I mean at this point it's either drop the column from the DB or change what's stored and display it in the report, so it seems like a change is warranted either way.  Got a preference?

snopoke replied:
> Got a preference?

Drop it ☠️

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -14,23 +18,46 @@ def make_url_key():
+    couch_user: CouchUser  # Is this correct?
+    request_method: str
+    user_agent: str
+    json: dict
```
nit: maybe call this 'body' or 'data'?

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -74,6 +74,12 @@ def to_context(self):
+@attr.s(kw_only=True, frozen=True, auto_attribs=True)
+class ApiResponse:
+    status: int
+    json: dict = None
```
note: if you rename the 'json' field in `RequestData` then rename this to match

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -90,3 +107,18 @@ def get_evaluation_context(restore_user, method, query, headers, body):
     })
+
+
+def retry(request_log):
```
nit: use a more verbose name: `reprocess_api_request`

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -14,22 +21,79 @@ def make_url_key():
-    except (UnicodeDecodeError, json.JSONDecodeError):
-        raise GenericInboundUserError(_("Payload must be valid JSON"))
+@attr.s(kw_only=True, frozen=True, auto_attribs=True)
+class RequestData:
```
nit: it would be nice to name this and `ApiResponse` similarly. Maybe `Api[Request|Response]Data`?

esoergel replied:
Agreed, I noticed that as well when writing this PR description - one of those things that's more obvious in retrospect.  Will do.

snopoke reviewed (commented):
Looking good so far. No major comments.

esoergel committed:
s/json/data/g

esoergel committed:
rename RequestData ApiRequest, like ApiResponse

I opted against ApiRequestData since it's meant to represent the entire
request, not just the data contained therein

esoergel committed:
Use more descriptive name

esoergel committed:
Require motech access to administer Generic API

At present any user on the domain can edit it, though only these users
will see the link

esoergel committed:
Remove error message from request log

esoergel committed:
Remove response_body field

I started just turning this into str(response.body) and adding a comment
like, "In the future...", but then I pictured myself looking at that
comment in a year and shaking my head about an imagined future that
never arrived, so I figured YAGNI should apply.  That said, if we're
fairly confident we _will_ want to add a response transform in the not
too distant future and we're confident that we'll log that as a string,
then keeping this in would let us avoid back-populating that field when
we get there.

snopoke reviewed (commented):
Looks good. Some minor stickler comments pending. Also this PR is still a draft.

esoergel committed:
stickler

esoergel committed:
Add retry test for invalid JSON (and assc fix)

esoergel committed:
Add view and button to call reprocess handler

esoergel commented on corehq/motech/generic_inbound/templates/generic_inbound/request_log_detail.html:
```diff
@@ -158,4 +158,15 @@ <h3 class="panel-title">{% trans "Processing Attempts" %}</h3>
 </div>
 
+<div class="form-actions">
+  {% if log.status != 'success' %}
```
I suppose I could also check this on the back-end, but that doesn't seem necessary

esoergel commented:
@snopoke I think this should be ready to go now, thanks for the patience

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -14,22 +21,78 @@ def make_url_key():
+@attr.s(kw_only=True, frozen=True, auto_attribs=True)
+class ApiRequest:
+    domain: str
+    couch_user: CouchUser  # Is this correct?
```
Is the comment still relevant?
```suggestion
    couch_user: CouchUser
```

snopoke commented on corehq/motech/generic_inbound/utils.py:
```diff
@@ -43,3 +106,21 @@ def get_evaluation_context(restore_user, method, query, headers, body):
+
+    with transaction.atomic():
+        request_log.status = RequestLog.Status.from_status_code(response.status)
+        request_log.response_status = response.status
```
This should also increment the attempt count.

snopoke reviewed (commented):
Few small comments

esoergel committed:
Remove outdated comment

Co-authored-by: Simon Kelly <skelly@dimagi.com>

esoergel committed:
Also update attempts when reprocessing
